lib/layer.js

```diff
@@ -41,10 +41,53 @@ function Layer (path, options, fn) {
   this.name = fn.name || '<anonymous>'
   this.params = undefined
   this.path = undefined
-  this.regexp = pathRegexp((opts.strict ? path : loosen(path)), this.keys, opts)
+  this.slash = path === '/' && opts.end === false
+
+  function matcher (_path) {
+    if (_path instanceof RegExp) {
+      var keys = []
+      var name = 0
+      var m
+      // eslint-disable-next-line no-cond-assign
+      while (m = MATCHING_GROUP_REGEXP.exec(_path.source)) {
+        keys.push({
+          name: m[1] || name++,
+          offset: m.index
+        })
+      }
+
+      return function regexpMatcher (p) {
+        var match = _path.exec(p)
+        if (!match) {
+          return false
+        }
+
+        var params = {}
+        for (var i = 1; i < match.length; i++) {
+          var key = keys[i - 1]
+          var prop = key.name
+          var val = decodeParam(match[i])
+
+          if (val !== undefined || !(hasOwnProperty.call(params, prop))) {
+            params[prop] = val
+          }
+        }
+
+        return {
+          params: params,
+          path: p
+        }
+      }
+    }
 
-  // set fast path flags
-  this.regexp._slash = path === '/' && opts.end === false
+    return pathRegexp.match((opts.strict ? _path : loosen(_path)), {
+      sensitive: opts.sensitive,
+      end: opts.end,
+      trailing: !opts.strict,
+      decode: decodeParam
+    })
+  }
+  this.matchers = Array.isArray(path) ? path.map(matcher) : [matcher(path)]
```

Comment:
I would personally recommend forwarding the array as-is through to `path-to-regexp` now, and only support non-array regexes. Mostly because I intent to work as much as possible on optimizing the matching code outside of router when possible. I.e. the path signature is `string | string[] | RegExp`.

Reply:
Oh I forgot we talked about that in slack but it was after I had added this here. I will remove that.